contracts/ring.md

```diff
@@ -0,0 +1,340 @@
+
+- Each contract defines a state type encapsulating its mutable state.
+- The value of the state is accessible from inside the contract
+  through an implicitly bound variable state.
+- State updates are performed by calling a function put : state =>
+  unit (possibly with a better name).
+- Aside from the put function (and similar functions for transactions
+  and events), the contract language is purely functional.
+- Functions modifying the state need to be annotated with the stateful keyword.
+
+To make it convenient to update parts of a deeply nested state Ring
+provides special special syntax for map/record updates.  Open
+question: we likely want to make it possible to have immutable state
+(parameters). Keep separate from mutable state or annotate certain
+fields as immutable?
+
+### Types
+Ring has the following types:
+
+| Type    | Description                     | Example
+| ------- | ------------------------------- | -------:
+| uint    | A 256 bit integer               | ```42```
+| int     | A 256 bit 2-complement integer  | ```-1```
+| address | A 256 bit number given as a hex | ```ff00```
+| bool    | A Boolean                       | ```true```
+| string  | An array of bytes               | ```“Foo”```
+| list    | A homogeneous immutable singly linked list. | ```[1, 2, 3]```
+| tuple   | An ordered heterogeneous array   | ```(42, “Foo”, true)```
+| record  | An immutable key value store with fixed key names and typed values | ``` type balance = { owner: address, value: uint } ```
+| map     | An immutable key value store with dynamic mapping of keys of one type to values of one type | ```type accounts = map(string, address)```
+| state   | A record of blockstate key, value pairs  |
+| transactions | An append only list of blockchain transactions |
+| events   | An append only list of blockchain events (or log entries) |
+
+#### Arrays
+Ring might not have arrays… Ring might have Maps.
+
+####  Variants
+Depending on how we write the compiler, if they are cheap to include
+some like to use variant types, while others might not want them.
+
+####  Refs
+ Ring has no refs.
+
+####  Object
+ Ring will not have objects. (only contracts) 
+
+### Pattern matching
+Pattern matching is probably outside the first iteration of Ring, but
+we definitely want it in the final language.
+
+### Builtins
+#### Events
+Ring has events which logs a structured message to the contract log in
+the resulting blockchain transaction.
+
+```
+event(name: string, value: any)
+```
+
+#### Transactions
+Ring can generate blockchain transactions.
+```
+transaction(tx_type: transaction, argument: record )
+```
+
+#### State
+
+A Ring contract has a state. You must declare the state record type in
+the contract. You can then update the state through mutable record
+update.
+
+```
+type state = { mutable field: string
+             , mutable value: uint};
+let state = {field: “”, value : 0};
+let store_value = (f: string, v: uint) => {
```

Comment:
This isn't compatible with the statement above that state is updated by calling put.

Reply:
Indeed. This section is left-over from a previous version of the spec.

Reply:
@UlfNorell Do you have a new version?

Reply:
The new version is already included at [the top](https://github.com/aeternity/protocol/blob/fa0c94983f6f4550d2de55ea7fa0c38ee2ab3507/contracts/ring.md#mutable-state).

Reply:
Hmm.
"State updates are performed by calling a function put : state =>
  unit (possibly with a better name)."
Does the function put only take a new state? Or does it take a field name and a value also?
If it only takes a state how do you create the new state syntactically? (as in the above description? how do you just update one of the fields?)
Some examples would be nice.

Reply:
There are some syntax example in the repo, but we should definitely describe it more precisely (and refine it). Briefly, we have syntax to create and update records:
```
  let newRec = {x: 42, y: "bar"};
  let updRec = newRec {x: 43};
```
It also allows deep updates
```
  let r2 = {r: updRec};
  let r2' = r2 {r.x: 44};
```
The parser doesn't yet handle it, but I want to use the same notation for maps, so you can write `m { [key]: val }` to set the value of `key` to `val` in the map `m`.

Also I think I prefer `=` over `:`, which is now possible since we don't have assignment.

So, to answer your question, you would write something like
```
  put(state { stakeholders[caller].balance: b })
```
to update a part of the state. There's some hope that the compiler would be able to compile this to an inplace memory write.